File: app/server/app/routers/reports.py
"""
Router para operaciones CRUD de Reports
"""

from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session
from typing import List, Optional
import uuid
from datetime import datetime

from app.database import get_db
from app.models import DailyReport as DailyReportModel
from app.schemas import DailyReport, DailyReportCreate

router = APIRouter()


@router.post("/", response_model=DailyReport)
def create_report(
    report: DailyReportCreate,
    db: Session = Depends(get_db)
):
    """Crear un nuevo reporte diario"""
    db_report = DailyReportModel(
        id=str(uuid.uuid4()),
        store_id=report.store_id,
        date=report.date,
        customers=report.customers,
        sales_cash=report.sales_cash,
        sales_card=report.sales_card,
        total_expenses=report.total_expenses,
        total_difference=report.total_difference,
        anomalies=report.anomalies,
        generated_at=datetime.now()
    )
    db.add(db_report)
    db.commit()
    db.refresh(db_report)
    return db_report


@router.get("/", response_model=List[DailyReport])
def get_reports(
    skip: int = 0,
    limit: int = 100,
    store_id: Optional[str] = None,
    date_from: Optional[str] = None,
    date_to: Optional[str] = None,
    db: Session = Depends(get_db)
):
    """Obtener todos los reportes diarios"""
    query = db.query(DailyReportModel)
    
    if store_id:
        query = query.filter(DailyReportModel.store_id == store_id)
    if date_from:
        query = query.filter(DailyReportModel.date >= date_from)
    if date_to:
        query = query.filter(DailyReportModel.date <= date_to)
    
    return query.order_by(DailyReportModel.date.desc()).offset(skip).limit(limit).all()


@router.get("/{report_id}", response_model=DailyReport)
def get_report(report_id: str, db: Session = Depends(get_db)):
    """Obtener un reporte por ID"""
    report = db.query(DailyReportModel).filter(
        DailyReportModel.id == report_id
    ).first()
    
    if not report:
        raise HTTPException(status_code=404, detail="Report not found")
    
    return report


@router.put("/{report_id}", response_model=DailyReport)
def update_report(
    report_id: str,
    report: DailyReportCreate,
    db: Session = Depends(get_db)
):
    """Actualizar un reporte"""
    db_report = db.query(DailyReportModel).filter(
        DailyReportModel.id == report_id
    ).first()
    
    if not db_report:
        raise HTTPException(status_code=404, detail="Report not found")
    
    for key, value in report.dict().items():
        setattr(db_report, key, value)
    
    db.commit()
    db.refresh(db_report)
    return db_report


@router.delete("/{report_id}")
def delete_report(report_id: str, db: Session = Depends(get_db)):
    """Eliminar un reporte"""
    report = db.query(DailyReportModel).filter(
        DailyReportModel.id == report_id
    ).first()
    
    if not report:
        raise HTTPException(status_code=404, detail="Report not found")
    
    db.delete(report)
    db.commit()
    return {"message": "Report deleted successfully"}

File: app/server/src/routes/reports.ts
import { Router, Request, Response } from 'express';
import { v4 as uuidv4 } from 'uuid';
import { Op } from 'sequelize';
import { DailyReport } from '../models/DailyReport';
import { dailyReportCreateSchema } from '../schemas/report';

const router = Router();

// Crear un nuevo reporte diario
router.post('/', async (req: Request, res: Response) => {
  try {
    const validatedData = dailyReportCreateSchema.parse(req.body);
    
    const report = await DailyReport.create({
      id: uuidv4(),
      ...validatedData,
      generated_at: new Date(),
    });

    res.status(201).json(report);
  } catch (error: any) {
    if (error.name === 'ZodError') {
      res.status(400).json({ error: 'Datos inválidos', details: error.errors });
    } else {
      res.status(500).json({ error: error.message });
    }
  }
});

// Obtener todos los reportes diarios
router.get('/', async (req: Request, res: Response) => {
  try {
    const { skip = 0, limit = 100, store_id, date_from, date_to } = req.query;
    
    const where: any = {};
    if (store_id) where.store_id = store_id;
    if (date_from && date_to) {
      where.date = {
        [Op.between]: [date_from, date_to],
      };
    } else if (date_from) {
      where.date = {
        [Op.gte]: date_from,
      };
    } else if (date_to) {
      where.date = {
        [Op.lte]: date_to,
      };
    }

    const reports = await DailyReport.findAll({
      where,
      limit: Number(limit),
      offset: Number(skip),
      order: [['date', 'DESC']],
    });

    res.json(reports);
  } catch (error: any) {
    res.status(500).json({ error: error.message });
  }
});

// Obtener un reporte por ID
router.get('/:report_id', async (req: Request, res: Response) => {
  try {
    const report = await DailyReport.findByPk(req.params.report_id);
    
    if (!report) {
      return res.status(404).json({ error: 'Report not found' });
    }

    res.json(report);
  } catch (error: any) {
    res.status(500).json({ error: error.message });
  }
});

// Actualizar un reporte
router.put('/:report_id', async (req: Request, res: Response) => {
  try {
    const validatedData = dailyReportCreateSchema.parse(req.body);
    const report = await DailyReport.findByPk(req.params.report_id);
    
    if (!report) {
      return res.status(404).json({ error: 'Report not found' });
    }

    await report.update(validatedData);
    res.json(report);
  } catch (error: any) {
    if (error.name === 'ZodError') {
      res.status(400).json({ error: 'Datos inválidos', details: error.errors });
    } else {
      res.status(500).json({ error: error.message });
    }
  }
});

// Eliminar un reporte
router.delete('/:report_id', async (req: Request, res: Response) => {
  try {
    const report = await DailyReport.findByPk(req.params.report_id);
    
    if (!report) {
      return res.status(404).json({ error: 'Report not found' });
    }

    await report.destroy();
    res.json({ message: 'Report deleted successfully' });
  } catch (error: any) {
    res.status(500).json({ error: error.message });
  }
});

export default router;

